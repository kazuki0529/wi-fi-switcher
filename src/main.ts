import { App } from '@aws-cdk/core';
import { WiFiSwitcherStack } from './wi-fi-switcher-stack';


// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new WiFiSwitcherStack(app, 'wi-fi-switcher-stack', { env: devEnv });

app.synth();
