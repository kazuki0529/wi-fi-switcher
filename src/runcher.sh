#!/bin/bash

set -eux

###########################################################
# 引数で渡された評価対象の日付が期間内か判定する
#
# @param target 評価対象の日付(yyyy-MM-dd hh:mm:ss)
# @param period_start 開始日(yyyy-MM-dd hh:mm:ss)
# @param period_end 終了日(yyyy-MM-dd hh:mm:ss)
#
# @return 期間内の場合は0を返し、期間外は1を返す
###########################################################
is_within_period() {
  declare -r target=$1
  [[ "${2}" = "" ]] \
    && declare -r period_start='1990-01-01 00:00:00' \
    || declare -r period_start=$2
  [[ "${3}" = "" ]] \
    && declare -r period_end='1990-01-01 00:00:00' \
    || declare -r period_end=$3

  # 評価対象が期間内か判定する
  declare -r start_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_start" +%s))
  declare -r end_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_end" +%s))
  if [ $start_difference -ge 0 ] && [ $end_difference -le 0 ]; then
    return 0;
  else
    return 1;
  fi
}

declare -r now=$1
declare -r schedule=$2

cd $(dirname $0)

while IFS=, read start end description; do
  # ヘッダ行判定
  [[ "${start}" == "start" ]] && continue

  is_within_period "${now}" "${start}" "${end}" \
    && (yarn cypress run -s ./cypress/integration/PR-500KI-wifi-on.ts || true) \
    &&  exit 0
done < "${schedule}"

yarn cypress run -s ./cypress/integration/PR-500KI-wifi-off.ts
