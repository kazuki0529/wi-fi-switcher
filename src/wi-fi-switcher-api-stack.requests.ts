import { APIGatewayEvent, APIGatewayProxyResult } from 'aws-lambda';
import moment from 'moment';
import RequestUseCase from './usecase/RequestUseCase';

export async function handler(
  event: APIGatewayEvent,
): Promise<APIGatewayProxyResult> {
  const usecase = RequestUseCase();

  const pathPart = event.resource.substring('/v1'.length);

  switch (pathPart) {
    case '/requests': {
      if (event.httpMethod === 'POST') {
        const body = event.body ? JSON.parse(event.body) : undefined;
        if (!body) throw new Error('The Body section is empty.');

        const req = {
          ...body,
          start: moment(body.start),
          end: moment(body.end),
        };

        const data = await usecase.create(req);
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data),
        };
      }
      if (event.httpMethod === 'GET') {
        const data = await usecase.getAll();
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data.map(e => ({
            ...e,
            start: e.start.toISOString(),
            end: e.end.toISOString(),
            createdAt: e.createdAt?.toISOString(),
            updatedAt: e.updatedAt?.toISOString(),
          }))),
        };
      }

      throw new Error('An undefined method has been called.');
    };
    case '/requests/{requestId}': {
      if (event.httpMethod !== 'PUT') throw new Error('An undefined method has been called.');

      const id = event.pathParameters ? event.pathParameters.requestId : undefined;
      if (!id) throw new Error('The id has not been specified.');

      const body = event.body ? JSON.parse(event.body) : undefined;
      if (!body) throw new Error('The Body section is empty.');

      const req = {
        ...body,
        start: body.start ? moment(body.start) : undefined,
        end: body.end ? moment(body.end) : undefined,
      };

      const data = await usecase.update(id, req);
      return {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json; charset=utf-8',
        },
        body: JSON.stringify(data),
      };
    };
    default: {
      throw new Error('An undefined path has been called.');
    };
  }
}
