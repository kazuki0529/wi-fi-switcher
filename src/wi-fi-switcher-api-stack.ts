import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamo from '@aws-cdk/aws-dynamodb';
import * as iam from '@aws-cdk/aws-iam';;
import * as lambda from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';

import { Construct, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';

export type StackStage = 'staging' | 'prod';

export class WiFiSwitcherApiStack extends Stack {
  public readonly table: dynamo.Table;
  public readonly api: apigateway.RestApi;

  constructor(
    scope: Construct, id: string,
    props?: StackProps,
  ) {
    super(scope, id, props);

    // データ格納領域作成
    this.table =new dynamo.Table(this, 'data', {
      partitionKey: {
        name: '_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_s_key',
        type: dynamo.AttributeType.STRING,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: dynamo.BillingMode.PAY_PER_REQUEST,
    });
    this.table.addGlobalSecondaryIndex({
      indexName: 'GSI-1',
      partitionKey: {
        name: '_gsi1_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_gsi1_s_key',
        type: dynamo.AttributeType.STRING,
      },
      projectionType: dynamo.ProjectionType.ALL,
    });

    /**
     * API作成
     */
    const requests = new NodejsFunction(this, 'requests', {
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_14_X,
      timeout: cdk.Duration.seconds(60),
      memorySize: 2048,
      environment: {
        REGION: this.region,
        TABLE_NAME: this.table.tableName,
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
      role: new iam.Role(this, 'generateBatchLambdaRole', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AWSLambdaBasicExecutionRole',
          ),
        ],
      }),
    });
    this.table.grantReadWriteData(requests);

    this.api = new apigateway.RestApi(this, 'game-play', {
      policy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            principals: [new iam.AnyPrincipal()],
            effect: iam.Effect.ALLOW,
            actions: ['execute-api:Invoke'],
            resources: ['execute-api:/prod/*'],
          }),
        ],
      }),
      deployOptions: {
        tracingEnabled: true,
        stageName: 'prod',
      },
    });
    const baseApi = this.api.root.addResource('v1');
    const requestsIntegration = new apigateway.LambdaIntegration(requests);
    const requestsApi = baseApi.addResource('requests');
    requestsApi.addMethod('POST', requestsIntegration);
    requestsApi.addMethod('GET', requestsIntegration);

    requestsApi.addResource('{requestId}').addMethod('PUT', requestsIntegration);

    new cdk.CfnOutput(this, 'ApiUrl', { value: this.api.url });
  }
}