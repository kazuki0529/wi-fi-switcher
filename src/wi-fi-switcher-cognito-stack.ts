import * as cognito from '@aws-cdk/aws-cognito';
import * as cdk from '@aws-cdk/core';

import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { StackStage } from './wi-fi-switcher-stack';

interface WiFiSwitcherCognitoStackProps extends StackProps {
  readonly stage: StackStage;
}

export class WiFiSwitcherCognitoStack extends Stack {
  readonly userPool: cognito.UserPool;
  readonly userPoolClient: cognito.UserPoolClient;

  constructor(
    scope: Construct, id: string,
    props: WiFiSwitcherCognitoStackProps,
  ) {
    super(scope, id, props);

    const zoneName = this.node.tryGetContext('ZONE_NAME');
    const domain = `${(props.stage === 'staging') ? 'dev.': ''}${zoneName}`;
    const fqdn = `game-play.home.${domain}`;

    this.userPool = new cognito.UserPool(this, 'userPool', {
      selfSignUpEnabled: false,
      standardAttributes: {
        email: { required: true, mutable: true },
        phoneNumber: { required: false },
      },
      signInCaseSensitive: false,
      autoVerify: { email: true },
      signInAliases: { username: true },
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    this.userPool.addDomain('domain', {
      cognitoDomain: { domainPrefix: fqdn.toLowerCase().split('.').join('-') },
    });

    this.userPoolClient = this.userPool.addClient('client', {
      oAuth: {
        scopes: [
          cognito.OAuthScope.EMAIL,
          cognito.OAuthScope.OPENID,
          cognito.OAuthScope.PROFILE,
        ],
        callbackUrls: [`https://${fqdn}/`],
        logoutUrls: [`https://${fqdn}/`],
        flows: { authorizationCodeGrant: true },
      },
    });
  }
}