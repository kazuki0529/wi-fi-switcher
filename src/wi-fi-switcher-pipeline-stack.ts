import * as codecommit from '@aws-cdk/aws-codecommit';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import { CodeCommitSourceAction } from '@aws-cdk/aws-codepipeline-actions';
import {
  Construct,
  Stage,
  Stack,
  StackProps,
  StageProps,
} from '@aws-cdk/core';
import {
  CdkPipeline,
  SimpleSynthAction,
} from '@aws-cdk/pipelines';
import {
  WiFiSwitcherApiStack,
} from './wi-fi-switcher-api-stack';
import {
  WiFiSwitcherStack,
} from './wi-fi-switcher-stack';


class Application extends Stage {
  constructor(
    scope: Construct,
    id: string,
    props?: StageProps,
  ) {
    super(scope, id, props);

    new WiFiSwitcherApiStack(
      this,
      'wi-fi-switcher-api',
    );
    new WiFiSwitcherStack(
      this,
      'wi-fi-switcher',
    );
  }
}

export class WiFiSwitcherPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const sourceArtifact = new codepipeline.Artifact();
    const cloudAssemblyArtifact = new codepipeline.Artifact();

    const repo = codecommit.Repository.fromRepositoryName(
      this,
      'repo',
      'wi-fi-switcher',
    ) as codecommit.Repository;

    const pipeline = new CdkPipeline(this, 'Pipeline', {
      pipelineName: 'Wi-Fi-SwitcherPipeline',
      cloudAssemblyArtifact,
      sourceAction: new CodeCommitSourceAction ({
        actionName: 'CodeCommit',
        repository: repo,
        branch: 'main',
        output: sourceArtifact,
      }),
      synthAction: SimpleSynthAction.standardYarnSynth({
        sourceArtifact,
        cloudAssemblyArtifact,
        environment: {
          privileged: true,
        },
        synthCommand: 'npx cdk synth',
      }),
    });

    // 開発用のDeploy
    pipeline.addApplicationStage(
      new Application(this, 'staging'),
    );

    // 本番用のDeploy
    pipeline.addApplicationStage(
      new Application(this, 'prod'),
      {
        manualApprovals: true,
      },
    );
  }
}
