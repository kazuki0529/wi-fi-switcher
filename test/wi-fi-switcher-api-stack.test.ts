import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { WiFiSwitcherApiStack } from '../src/wi-fi-switcher-api-stack';

test('Snapshot', () => {
  const app = new App();
  const stack = new WiFiSwitcherApiStack(app, 'test');

  expect(stack).toHaveResource('AWS::DynamoDB::Table');

  expect(stack).toHaveResource('AWS::ApiGateway::RestApi');
  expect(stack).toHaveResource('AWS::Lambda::Function');

  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});
