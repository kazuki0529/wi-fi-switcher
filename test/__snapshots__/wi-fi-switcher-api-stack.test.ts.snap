// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927ArtifactHashDA81D15F": Object {
      "Description": "Artifact hash for asset \\"04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927\\"",
      "Type": "String",
    },
    "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927S3Bucket86ACEF8B": Object {
      "Description": "S3 bucket for asset \\"04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927\\"",
      "Type": "String",
    },
    "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927S3VersionKey59C87B76": Object {
      "Description": "S3 key for asset version \\"04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927\\"",
      "Type": "String",
    },
    "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686ArtifactHashB7ECE5F1": Object {
      "Description": "Artifact hash for asset \\"c8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686\\"",
      "Type": "String",
    },
    "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686S3Bucket821B928E": Object {
      "Description": "S3 bucket for asset \\"c8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686\\"",
      "Type": "String",
    },
    "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686S3VersionKey8D940F9A": Object {
      "Description": "S3 key for asset version \\"c8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686S3Bucket821B928E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686S3VersionKey8D940F9A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc8dfcc3db7a767ebc6f9417372a8fee940927c659f793315834a6784dab3d686S3VersionKey8D940F9A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "data7E2128CA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "_p_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "_s_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "_gsi1_p_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "_gsi1_s_key",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "GSI-1",
            "KeySchema": Array [
              Object {
                "AttributeName": "_gsi1_p_key",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "_gsi1_s_key",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "_p_key",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "_s_key",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "gameplayCFF713D3": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowHeaders": Array [
            "content-type",
            "authorization",
            "x-amz-date",
            "x-api-key",
            "x-amz-security-token",
            "x-amz-user-agent",
          ],
          "AllowMethods": Array [
            "*",
          ],
          "AllowOrigins": Array [
            "*",
          ],
        },
        "Name": "game-play",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "gameplayDefaultStageE4871BAA": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "gameplayGETv1requestsE1DFF4F4": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "gameplayauthorizerB5BCBBEB",
        },
        "RouteKey": "GET /v1/requests",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "gameplayGETv1requestsapirequestintegration8AB6D9A4",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "gameplayGETv1requestsapirequestintegration8AB6D9A4": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "requestsC775BB57",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "gameplayGETv1requestsapirequestintegrationPermission7AAA04F4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "requestsC775BB57",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "gameplayCFF713D3",
              },
              "/*/*/v1/requests",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "gameplayPOSTv1requests744637FD": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "gameplayauthorizerB5BCBBEB",
        },
        "RouteKey": "POST /v1/requests",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "gameplayGETv1requestsapirequestintegration8AB6D9A4",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "gameplayPOSTv1requestsapirequestintegrationPermission76C46491": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "requestsC775BB57",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "gameplayCFF713D3",
              },
              "/*/*/v1/requests",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "gameplayPUTv1requestsrequestId932050D1": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "gameplayauthorizerB5BCBBEB",
        },
        "RouteKey": "PUT /v1/requests/{requestId}",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "gameplayPUTv1requestsrequestIdapirequestintegrationF1852542",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "gameplayPUTv1requestsrequestIdapirequestintegrationF1852542": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "requestsC775BB57",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "gameplayPUTv1requestsrequestIdapirequestintegrationPermission9177BDC7": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "requestsC775BB57",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "gameplayCFF713D3",
              },
              "/*/*/v1/requests/{requestId}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "gameplayauthorizerB5BCBBEB": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "gameplayCFF713D3",
        },
        "AuthorizerType": "JWT",
        "IdentitySource": Array [
          "$request.header.Authorization",
        ],
        "JwtConfiguration": Object {
          "Audience": Array [
            Object {
              "Fn::ImportValue": "cognito:ExportsOutputRefuserPoolclient992D31C6EE176468",
            },
          ],
          "Issuer": Object {
            "Fn::Join": Array [
              "",
              Array [
                "https://cognito-idp.",
                Object {
                  "Ref": "AWS::Region",
                },
                ".amazonaws.com/",
                Object {
                  "Fn::ImportValue": "cognito:ExportsOutputRefuserPoolDC9497E0A5AAB3BA",
                },
              ],
            ],
          },
        },
        "Name": "authorizer",
      },
      "Type": "AWS::ApiGatewayV2::Authorizer",
    },
    "generateBatchLambdaRole0660E91E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "generateBatchLambdaRoleDefaultPolicy2B30E671": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "data7E2128CA",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "data7E2128CA",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "generateBatchLambdaRoleDefaultPolicy2B30E671",
        "Roles": Array [
          Object {
            "Ref": "generateBatchLambdaRole0660E91E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "requestsC775BB57": Object {
      "DependsOn": Array [
        "generateBatchLambdaRoleDefaultPolicy2B30E671",
        "generateBatchLambdaRole0660E91E",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927S3Bucket86ACEF8B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927S3VersionKey59C87B76",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters04bf825a5ceb5b5f28507cdd99c42b45d8d66699ee3bd380206a156f52d66927S3VersionKey59C87B76",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "REGION": Object {
              "Ref": "AWS::Region",
            },
            "TABLE_NAME": Object {
              "Ref": "data7E2128CA",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 2048,
        "Role": Object {
          "Fn::GetAtt": Array [
            "generateBatchLambdaRole0660E91E",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "requestsLogRetention6E8BFD33": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "requestsC775BB57",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
  },
}
`;
